tools/HogEdit/HogEdit.cpp

```diff
@@ -0,0 +1,339 @@
+  return std::chrono::system_clock::to_time_t(sctp);
+}
+
+D3::HogFormat::hog_iter find_file_entry(const D3::HogFormat &format, const std::string &name) {
+  auto fn = [&name](const auto &hfe) { return D3::lowercase(name) == D3::lowercase(D3::arr_to_str(hfe.name)); };
+  if (auto it = std::find_if(std::begin(format), std::end(format), fn); it != std::end(format))
+    return it;
+  return std::end(format);
+}
+
+std::list<std::filesystem::path> validate_files(const std::filesystem::path &base_path,
+                                                std::list<std::filesystem::path> filepaths) {
+  bool printed_base_dir = false;
+  std::list<std::filesystem::path> valid_files;
+  for (auto &filepath : filepaths) {
+    if (!std::filesystem::exists(base_path / filepath)) {
+      if (!printed_base_dir) {
+        printed_base_dir = true;
+        std::cout << "File paths relative to " << std::filesystem::absolute(base_path) << std::endl;
+      }
+      std::cout << "File " << filepath << " does not exist! Skipping... " << std::endl;
+    } else {
+      if (!std::filesystem::is_regular_file(base_path / filepath))
+        throw std::runtime_error(filepath.string() + " is not a regular file.");
+
+      if (filepath.string().size() > 36)
+        throw std::runtime_error("Length of filepath \"" + filepath.string() + "\" exceeds 36 bytes.");
+
+      valid_files.emplace_back(filepath);
+    }
+  }
+  return valid_files;
+}
+
+namespace D3 {
+std::filesystem::path archive_path; // path of hog archive
+std::filesystem::path base_path;
+std::list<std::filesystem::path> file_path_list;
+
+void add_files_to_header(D3::HogFormat &format, const std::list<std::filesystem::path> &path_list) {
+  for (const auto &i : path_list) {
+    D3::HogFileEntry entry;
+    auto name = i.u8string();
+    std::copy(name.begin(), name.end(), entry.name.begin());
+    entry.flags = 0;
+    entry.len = static_cast<uint32_t>(std::filesystem::file_size(base_path / i));
+    entry.timestamp = static_cast<uint32_t>(file_time_to_uint32_t(std::filesystem::last_write_time(base_path / i)));
+
+    format.AddEntry(entry);
+  }
+}
+
+void amend_hog(D3::HogFormat &old_header, D3::HogFormat &new_header) {
+  std::ifstream old_hog;
+  std::ofstream new_hog;
+  auto tmp_archive = archive_path;
+  tmp_archive += ".tmp";
+  bool modify_mode = std::distance(std::begin(old_header), std::end(old_header));
+
+  if (modify_mode) {
+    old_hog = std::ifstream(archive_path, std::ios_base::in | std::ios_base::binary);
+    new_hog = std::ofstream(tmp_archive, std::ios_base::out | std::ios_base::binary | std::ios_base::trunc);
+  } else
+    new_hog = std::ofstream(archive_path, std::ios_base::out | std::ios_base::binary);
+
+  new_hog << new_header;
+
+  for (const D3::HogFileEntry &entry : new_header) {
+    if (auto it = find_file_entry(old_header, arr_to_str(entry.name)); it != std::end(old_header)) {
+      std::cout << "Copying " << arr_to_str(entry.name) << "... ";
```

Comment:
Adding a method which returns an `std::string_view` view into `HogFileEntry::name` could replace `arr_to_str()`. However then the case-insensitive comparison below needs to be adapted.